gcsio/src/main/java/com/google/cloud/hadoop/gcsio/InvocationIdGrpcChannelInterceptor.java

```diff
@@ -0,0 +1,57 @@
+package com.google.cloud.hadoop.gcsio;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import java.util.UUID;
+
+public final class InvocationIdGrpcChannelInterceptor implements ClientInterceptor {
+
+  static final Metadata.Key<String> GOOG_API_CLIENT =
+      Metadata.Key.of("x-goog-api-client", Metadata.ASCII_STRING_MARSHALLER);
+
+  static final String GCCL_INVOCATION_ID = "gccl-invocation-id";
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    return new InvocationIdAttachingClientCall<>(next.newCall(method, callOptions));
+  }
+
+  // Non private to avoid synthetic class
+  InvocationIdGrpcChannelInterceptor() {}
+
+  private final class InvocationIdAttachingClientCall<ReqT, RespT>
+      extends SimpleForwardingClientCall<ReqT, RespT> {
+
+    // Non private to avoid synthetic class
+    InvocationIdAttachingClientCall(ClientCall<ReqT, RespT> call) {
+      super(call);
+    }
+
+    @Override
+    public void start(Listener<RespT> responseListener, Metadata headers) {
+      appendInvocationId(headers);
+      super.start(responseListener, headers);
+    }
+
+    private void appendInvocationId(final Metadata header) {
+      UUID uuid = UUID.randomUUID();
+      String invocationId = String.format("%s/%s", GCCL_INVOCATION_ID, uuid.toString());
+      final String value = header.get(GOOG_API_CLIENT);
+      if (value != null) {
+        // First remove the value from Metadata, as overrides are not supported
+        header.remove(GOOG_API_CLIENT, value);
+        // append space separated invocationId
+        String appendedValue = value + " " + invocationId;
+        header.put(GOOG_API_CLIENT, appendedValue);
```

Comment:
Is this the recommended approach? I did not see the key getting overwritten, from the implementation.

Also what will happen if another interceptor "puts" this header.

Reply:
That is correct keys are not overwritten. That's what I commented too. Metadata don't verify if key is already there or not and just appends it at the end of an array of headers.

Just trying to be in sync with how this is being done as library (manual client) where we are appending it to existing header instead of duplicating the key.

Reply:
For this to work the "value" in the header cannot have  space correct? I am not sure if there is any such restrictions.